### PR TITLE
add support for OpenSearch vector store

### PR DESCRIPTION
**Pull request by JM-Lab**
Add OpenSearchAiSearchFilterExpressionConverter
Add support for FilterExpression
Add support for OpenSearch vector store

### CODE REVIEW

**Comment by eddumelendez** [comment]
Hi @JM-Lab, AWS also provides OpenSearch and it the java sdk provides [AwsSdk2Transport](https://github.com/opensearch-project/opensearch-java/blob/ef04bf87b3557e6f090e41859a8c356a111224e2/java-client/src/main/java/org/opensearch/client/transport/aws/AwsSdk2Transport.java). Wonder if should be part of this PR or a separate one.

**Comment by JM-Lab** [comment]
Hi @eddumelendez,

Sure, use AwsSdk2Transport for OpenSearchClient. Check this link for details: https://opensearch.org/docs/latest/clients/java/#connecting-to-amazon-opensearch-service

File: vector-stores/spring-ai-opensearch-store/pom.xml [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+		<dependency>
+			<groupId>org.springframework.ai</groupId>
+			<artifactId>spring-ai-core</artifactId>
+			<version>${parent.version}</version>
+		</dependency>
+
+		<dependency>
+			<groupId>org.opensearch.client</groupId>
+			<artifactId>opensearch-java</artifactId>
+			<version>2.9.1</version>
```
Comment (tzolov) [inline vector-stores/spring-ai-opensearch-store/pom.xml:38]: Consider moving the version as a property in the parent POM

File: vector-stores/spring-ai-opensearch-store/pom.xml [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+		 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+	<modelVersion>4.0.0</modelVersion>
+	<parent>
+		<groupId>org.springframework.ai</groupId>
+		<artifactId>spring-ai</artifactId>
+		<version>1.0.0-SNAPSHOT</version>
+		<relativePath>../../pom.xml</relativePath>
+	</parent>
+	<artifactId>spring-ai-opensearch-store</artifactId>
```
Comment (tzolov) [inline vector-stores/spring-ai-opensearch-store/pom.xml:11]: And the spring-ai-opensearch-store dependency to the spring-bom

File: vector-stores/spring-ai-opensearch-store/pom.xml [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+		</dependency>
+
+		<dependency>
+			<groupId>org.opensearch.client</groupId>
+			<artifactId>opensearch-java</artifactId>
+			<version>2.9.1</version>
+		</dependency>
+
+		<dependency>
+			<groupId>org.apache.httpcomponents.client5</groupId>
```
Comment (tzolov) [inline vector-stores/spring-ai-opensearch-store/pom.xml:42]: Is this a test dependency only?

File: vector-stores/spring-ai-opensearch-store/src/main/java/org/springframework/ai/vectorstore/OpenSearchVectorStore.java [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+        try {
+            BooleanResponse response = this.openSearchClient.indices()
+                    .exists(existRequestBuilder -> existRequestBuilder.index(targetIndex));
+            return response.value();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public CreateIndexResponse createIndexMapping(String index, Map<String, Property> properties) {
```
Comment (tzolov) [inline vector-stores/spring-ai-opensearch-store/src/main/java/org/springframework/ai/vectorstore/OpenSearchVectorStore.java:186]: is this supposed to be public?

**Review by tzolov — CHANGES_REQUESTED** [review]
Thank you @JM-Lab , 
You get an extra points for implementing the filter expressions as well ;)
I made few comments inline. In addition would you consider:
- Adding documentation under the [vectordbs](https://github.com/spring-projects/spring-ai/tree/main/spring-ai-docs/src/main/antora/modules/ROOT/pages/api/vectordbs) and add the references to the catalog nav.adoc.
- Add opensearch auto-configuration and boot starter

File: vector-stores/spring-ai-opensearch-store/pom.xml (hunk 3 above)
Comment (JM-Lab) [reply vector-stores/spring-ai-opensearch-store/pom.xml:42]: It's a required dependency for OpenSearch client. See https://opensearch.org/docs/latest/clients/java/#installing-the-client-using-apache-httpclient-5-transport.

File: vector-stores/spring-ai-opensearch-store/src/main/java/org/springframework/ai/vectorstore/OpenSearchVectorStore.java (hunk 4 above)
Comment (JM-Lab) [reply vector-stores/spring-ai-opensearch-store/src/main/java/org/springframework/ai/vectorstore/OpenSearchVectorStore.java:186]: Yes, It's public to handle different sizes for each embedding model and allow for custom metadata mappings.

**Comment by JM-Lab** [comment]
Hi @tzolov,

I plan to address your requests within the next 1-2 weeks. Would you prefer me to add commits to the current pull request or handle them in a separate pull request? Let me know what you think.

File: vector-stores/spring-ai-opensearch-store/src/main/java/org/springframework/ai/vectorstore/OpenSearchVectorStore.java (hunk 4 above)
Comment (JM-Lab) [reply vector-stores/spring-ai-opensearch-store/src/main/java/org/springframework/ai/vectorstore/OpenSearchVectorStore.java]: https://github.com/spring-projects/spring-ai/pull/508/commits/7b77b193c0bc4cc54abf49041b932725ca268568

Refer to the commit where I added the mappingJson parameter to OpenSearchVectorStore to handle different embedding sizes and allow custom metadata mappings.

**Comment by JM-Lab** [comment]
- [ ] Adding documentation under the [vectordbs](https://github.com/spring-projects/spring-ai/tree/main/spring-ai-docs/src/main/antora/modules/ROOT/pages/api/vectordbs)
- [ ] add the references to the catalog nav.adoc.
Add opensearch auto-configuration and boot starter

**Comment by JM-Lab** [comment]
Hi @tzolov,

Will reset tangled commits and soon add documentation for vectordbs with updates to catalog nav.adoc for a clean pull request.

**Comment by JM-Lab** [comment]
@tzolov

I've just created a new PR focusing solely on the OpenSearch files, on the main branch. You can find it here: https://github.com/spring-projects/spring-ai/pull/663.

Going to close this current one. Thanks!

**Comment by JM-Lab** [comment]
@tzolov

I've just created a new PR focusing solely on the OpenSearch files, on the main branch. You can find it here: https://github.com/spring-projects/spring-ai/pull/663.

Going to close this current one. Thanks!

**Pull request closed without merging**